File: apps/web/src/api/policies.ts
import { apiFetch } from "../lib/fetch";
import { PolicyIn, PolicyOut, DryRunResult } from "../schemas/policy";
import { z } from "zod";

export async function listPolicies() {
  const data = await apiFetch(`/policies`);
  return z.object({ policies: z.array(PolicyOut) }).parse(data);
}

export async function applyPolicy(
  policy: z.infer<typeof PolicyIn>,
  dryRunOnly?: boolean,
) {
  const qs = new URLSearchParams();
  if (dryRunOnly) qs.set("dry_run_only", "true");
  const data = await apiFetch(`/policies/apply?${qs.toString()}`, {
    method: "POST",
    body: JSON.stringify(policy),
  });
  try {
    return DryRunResult.parse(data);
  } catch {
    return data as { id: string; version: number };
  }
}

File: apps/api/routers/policies.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from apps.api.core.db import get_session
from apps.api.schemas.policy import PolicyIn
from apps.api.services.policy_service import dry_run, list_policies, upsert_policy

router = APIRouter()


@router.get("/policies")
def get_policies(db: Session = Depends(get_session)):
    rows = list_policies(db)
    return {
        "policies": [
            {
                "id": p.id,
                "name": p.name,
                "match": p.match,
                "rules": p.rules,
                "version": p.version,
            }
            for p in rows
        ]
    }


@router.post("/policies/apply")
def apply_policy(
    body: PolicyIn, dry_run_only: bool = False, db: Session = Depends(get_session)
):
    if dry_run_only:
        return dry_run(db, body)
    p = upsert_policy(db, body)
    return {"id": p.id, "version": p.version}
